app/create-pr.js

```diff
@@ -0,0 +1,171 @@
+ * read/write permissions for contents and pull requests
+ */
+const auth = process.env.GITHUB_API_KEY;
+const owner = "vincerubinetti";
+const repo = "captions";
+
+/** endpoint */
+functions.http("create-pr", async (request, response) => {
+  response.set("Access-Control-Allow-Origin", "*");
+  response.set("Access-Control-Allow-Methods", "POST");
+  response.set("Access-Control-Allow-Headers", "Accept, Content-Type");
+
+  if (request.method == "OPTIONS") return response.status(200).send();
+
+  try {
+    const pr = await createPr(request.body);
+    return response.status(200).json(pr);
+  } catch (error) {
+    return response.status(400).send(error.message);
+  }
+});
+
+/** create/update branch, file, pr */
+async function createPr(params, debug = false) {
+  /** auth with github */
+  const octokit = new Octokit({ auth });
+
+  /** get params */
+  const { branch, title, body, files = [] } = params || {};
+
+  /** check for missing params */
+  const missing = [];
+  if (!branch) missing.push("branch");
+  if (!title) missing.push("title");
+  if (!body) missing.push("body");
+  if (!files.length) missing.push("files");
+  if (missing.length) throw Error(`Insufficient ${missing.join(", ")}`);
+
+  /** get main branch */
+  let main;
+  try {
+    main = (
+      await octokit.rest.git.getRef({
+        owner,
+        repo,
+        ref: "heads/main",
+      })
+    ).data.object.sha;
+  } catch (error) {
+    if (debug) console.warn(error);
+    throw Error(`Couldn't get main branch: ${error?.response?.data?.message}`);
+  }
+
+  /** get existing branch */
+  let newBranch;
+  try {
+    newBranch = await octokit.rest.git.getRef({
+      owner,
+      repo,
+      ref: `heads/${branch}`,
+    });
+  } catch (error) {
+    if (debug) console.warn(error);
+    console.warn(
+      `Couldn't get existing branch ${branch}: ${error?.response?.data?.message}`,
+    );
+  }
+
+  if (!newBranch)
+    /** create new branch */
+    try {
+      newBranch = await octokit.rest.git.createRef({
+        owner,
+        repo,
+        ref: `refs/heads/${branch}`,
+        sha: main,
+      });
+    } catch (error) {
+      if (debug) console.warn(error);
+      throw Error(
+        `Couldn't create new branch ${branch}: ${error?.response?.data?.message}`,
+      );
+    }
+
+  /** update files */
+  for (const [index, { path, content }] of Object.entries(files)) {
+    if (!path || !content) throw Error(`File ${index} insufficient`);
+
+    /** get existing file */
+    let existing;
+    try {
+      existing = (
+        await octokit.rest.repos.getContent({
+          owner,
+          repo,
+          branch,
+          path,
+        })
+      ).data.sha;
+    } catch (error) {
+      if (debug) console.warn(error);
+      console.warn(
+        `Couldn't get existing file ${path}: ${error?.response?.data?.message}`,
+      );
+    }
+
+    /** create or update file */
+    try {
+      await octokit.rest.repos.createOrUpdateFileContents({
```

Comment:
As written, this just overwrites the whole file with a new one, which is just generated from the title/captions/description state in the app. See `revert()` in `/src/data/data.ts`; maybe we need to re-include things like start/end time and model there? (In testing, I saw some sentence_translation.json files that still had start/end times, but I thought we moved those to just the english folder?)

Reply:
I ended up moving the start and end times back to the translation files since as it stands the sentences do not always line up (due to two different runs of Whisper). In principle, it's possible to deduce from the word timings, but in favor of a simpler app, I'm okay with the added data/redundancy.